### PR TITLE
Complete EVM -> Cardano Relayer works

### PR DESCRIPTION
**Pull request by hai-kreate**

### CODE REVIEW

File: rust/chains/hyperlane-cardano/Cargo.toml [hunk 1]
```diff
@@ -15,4 +15,4 @@ hex = "0.4.3"
 tracing = { workspace = true }
 reqwest = { workspace = true }
 tokio = { workspace = true, features = ["macros"] }
-hyperlane-cardano-rpc-rust-client = "1.0.2"
\ No newline at end of file
+hyperlane-cardano-rpc-rust-client = { path = "../../../../hyperlane-cardano/generated-sources/hyperlane-cardano-rpc-rust-client" }
```
Comment (hai-kreate) [inline rust/chains/hyperlane-cardano/Cargo.toml]: This is only development. Should publish from `hyperlane-cardano` before reverting and merging this.

File: rust/chains/hyperlane-cardano/src/multisig_ism.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+impl MultisigIsm for CardanoMultisigIsm {
+    /// Returns the validator and threshold needed to verify message
+    async fn validators_and_threshold(
+        &self,
+        message: &HyperlaneMessage,
+    ) -> ChainResult<(Vec<H256>, u8)> {
+        // We're using the same multisig ISM for all messages
+        // TODO: https://github.com/tvl-labs/hyperlane-cardano/issues/42
+        // will enable dApp-defined ISM
+        let parameters = self.cardano_rpc.get_ism_parameters().await.unwrap();
```
Comment (distributedstatemachine) [inline rust/chains/hyperlane-cardano/src/multisig_ism.rs:59]: Can we handle this error and not make it panic?

File: rust/chains/hyperlane-cardano/src/multisig_ism.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        let validators = parameters
+            .validators
+            .iter()
+            .map(|v| H256::from_str(v).unwrap())
```
Comment (distributedstatemachine) [inline rust/chains/hyperlane-cardano/src/multisig_ism.rs:63]: Same here

File: rust/chains/hyperlane-cardano/src/multisig_ism.rs (hunk 2 above)
Comment (hai-kreate) [reply rust/chains/hyperlane-cardano/src/multisig_ism.rs:59]: We can do it after the MVP for the whole app, currently it's the norm everywhere:
- https://github.com/tvl-labs/hyperlane-monorepo/pull/3/files#diff-f19dbfe4182f95fe54e033dc8e1a6d970807c66b1b4f22208c268673e555c73fR62
- https://github.com/tvl-labs/hyperlane-cardano/issues/59

File: rust/chains/hyperlane-cardano/src/rpc/mod.rs [hunk 4]
```diff
@@ -51,6 +55,100 @@ impl CardanoRpc {
...
+            },
+        )
+        .await
+    }
+
+    // TODO: Share metadata decoding code with estimating fee
+    pub async fn submit_inbox_message(
+        &self,
+        message: &HyperlaneMessage,
+        metadata: &[u8],
```
Comment (serejke) [inline rust/chains/hyperlane-cardano/src/rpc/mod.rs:146]: We can introduce a `CardanoMessageMetadata` struct with encode/decode code

File: rust/chains/hyperlane-cardano/src/rpc/mod.rs [hunk 5]
```diff
@@ -51,6 +55,100 @@ impl CardanoRpc {
...
+        submit_inbound_message(
+            &self.0,
+            SubmitInboundMessageRequest {
+                // TODO: Read from config
+                relayer_cardano_address:
+                    "addr_test1vqvjvk3qezccu5a3gce65mqvg4tpfy47plv68wmh68paswqv3jaqe".to_string(),
+                // TODO: Read from config
+                private_key: "e8e34f6c74e22577d609803dfe9c8773f10e478e7dadf6d065a78ae42a21f912"
+                    .to_string(),
+                origin: message.origin as f32,
```
Comment (serejke) [inline rust/chains/hyperlane-cardano/src/rpc/mod.rs]: We need to update RPC OpenAPI to have `integer` type instead of `number` to avoid these floating point types.

File: rust/agents/relayer/src/msg/metadata/multisig/base.rs [hunk 6]
```diff
@@ -153,8 +153,10 @@ fn order_signatures(desired_order: &[H256], signatures: &[SignatureWithSigner])
         .iter()
         .cloned()
         .map(|s| {
-            let order_index = ordering_map.get(&H256::from(s.signer)).unwrap();
-            (s, *order_index)
+            // TODO: Revert this
```
Comment (serejke) [inline rust/agents/relayer/src/msg/metadata/multisig/base.rs]: Why is this necessary? Can we revert now? Is there a task?

File: rust/agents/relayer/src/msg/metadata/multisig/message_id_multisig.rs [hunk 7]
```diff
@@ -34,8 +38,9 @@ impl MultisigIsmMetadataBuilder for MessageIdMultisigMetadataBuilder {
         checkpoint_syncer: &MultisigCheckpointSyncer,
     ) -> Result<Option<MultisigMetadata>> {
         const CTX: &str = "When fetching MessageIdMultisig metadata";
+        // TODO: Select checkpoint depending on destination chain
```
Comment (serejke) [inline rust/agents/relayer/src/msg/metadata/multisig/message_id_multisig.rs]: We need this be done now, to use the same agent Docker image for both EVM -> Cardano and Cardano -> EVM flows.

Should be straightforward: https://github.com/tvl-labs/hyperlane-monorepo/blob/93fb6ffcd39adafbed4d63ee776e0cbef82610cf/rust/hyperlane-core/src/types/message.rs#L168

File: rust/hyperlane-base/src/types/multisig.rs [hunk 8]
```diff
@@ -358,4 +358,64 @@ impl MultisigCheckpointSyncer {
...
+                        debug!(
+                            validator = format!("{:#x}", validator),
+                            index = index,
+                            checkpoint_index = signed_checkpoint.value.index,
+                            "Checkpoint index mismatch"
+                        );
+                        continue;
+                    }
+
+                    // TODO: Ensure that the signature is actually by the validator
```
Comment (serejke) [inline rust/hyperlane-base/src/types/multisig.rs]: Please implement these TODOs in this PR

File: rust/hyperlane-base/src/types/multisig.rs [hunk 9]
```diff
@@ -358,4 +358,64 @@ impl MultisigCheckpointSyncer {
...
+                    // TODO: Support multisig checks
+                    // For Cardano/Blake2b we don't recover
+                    let signer = H160::zero(); // TODO: Get address from public key
```
Comment (serejke) [inline rust/hyperlane-base/src/types/multisig.rs]: Please also implement this TODO

File: rust/agents/validator/src/submit.rs [hunk 10]
```diff
@@ -125,12 +136,29 @@ impl ValidatorSubmitter {
...
-                        let signed_checkpoint = self
-                            .signer
-                            .sign(CheckpointWithMessageIdBlake2b {
-                                checkpoint: queued_checkpoint,
-                            })
-                            .await?;
+                        let checkpoint_blake2b = CheckpointWithMessageIdBlake2b {
+                            checkpoint: queued_checkpoint,
+                        };
+                        let secret_key =
```
Comment (serejke) [inline rust/agents/validator/src/submit.rs]: We cat try to extend `Signer` trait with a our own function `sign_raw` that eventually calls non EVM sign on ethers' AWS (or Signers::Local) provider.

I think we'd better have a bit more code that one more extremely important configuration parameter (`secretKey`).

**Review by serejke — CHANGES_REQUESTED** [review]

File: rust/chains/hyperlane-cardano/src/rpc/mod.rs (hunk 4 above)
Comment (hai-kreate) [reply rust/chains/hyperlane-cardano/src/rpc/mod.rs:146]: Sure, will do.

File: rust/chains/hyperlane-cardano/src/rpc/mod.rs (hunk 5 above)
Comment (hai-kreate) [reply rust/chains/hyperlane-cardano/src/rpc/mod.rs]: That's a lot better, yes!

File: rust/agents/relayer/src/msg/metadata/multisig/base.rs (hunk 6 above)
Comment (hai-kreate) [reply rust/agents/relayer/src/msg/metadata/multisig/base.rs]: It's a shortcut to compile the agent tracked here:
https://github.com/tvl-labs/hyperlane-cardano/issues/70

I can fix it in this PR.

File: rust/agents/relayer/src/msg/metadata/multisig/message_id_multisig.rs (hunk 7 above)
Comment (hai-kreate) [reply rust/agents/relayer/src/msg/metadata/multisig/message_id_multisig.rs]: 100%, will fix.

File: rust/hyperlane-base/src/types/multisig.rs (hunk 8 above)
Comment (hai-kreate) [reply rust/hyperlane-base/src/types/multisig.rs]: Okay.

File: rust/hyperlane-base/src/types/multisig.rs (hunk 9 above)
Comment (hai-kreate) [reply rust/hyperlane-base/src/types/multisig.rs]: Okay.

File: rust/agents/validator/src/submit.rs (hunk 10 above)
Comment (hai-kreate) [reply rust/agents/validator/src/submit.rs]: Can an AWS signer sign raw or return the private key for us to sign raw?

Comment (serejke) [reply rust/agents/validator/src/submit.rs]: Yes, 
https://github.com/tvl-labs/hyperlane-monorepo/blob/93fb6ffcd39adafbed4d63ee776e0cbef82610cf/rust/chains/hyperlane-ethereum/src/signers.rs#L43

The current code calls `sign_message`, which internally calls `AwsSigner.sign_digest` (raw)

File: rust/agents/relayer/src/msg/metadata/multisig/message_id_multisig.rs [hunk 11]
```diff
@@ -34,28 +38,58 @@ impl MultisigIsmMetadataBuilder for MessageIdMultisigMetadataBuilder {
         checkpoint_syncer: &MultisigCheckpointSyncer,
     ) -> Result<Option<MultisigMetadata>> {
         const CTX: &str = "When fetching MessageIdMultisig metadata";
-        let Some(quorum_checkpoint) = checkpoint_syncer
+        match KnownHyperlaneDomain::try_from(message.destination) {
+            Ok(KnownHyperlaneDomain::CardanoTest1) => {
+                // TODO: Deduplicate code between keccak checkpoint & blake2b checkpoint types
```
Comment (serejke) [inline rust/agents/relayer/src/msg/metadata/multisig/message_id_multisig.rs]: Could we use `TODO[cardano]: <text>`? I believe it will be easier to filter TODOs by Cardano-related ones

File: rust/hyperlane-base/src/types/multisig.rs [hunk 12]
```diff
@@ -358,4 +358,71 @@ impl MultisigCheckpointSyncer {
...
+                            index = index,
+                            checkpoint_index = signed_checkpoint.value.index,
+                            "Checkpoint index mismatch"
+                        );
+                        continue;
+                    }
+
+                    // TODO: Support multisig checks
+                    let signer = signed_checkpoint.recover_raw()?;
+                    println!("Signer {:?}", signer);
```
Comment (serejke) [inline rust/hyperlane-base/src/types/multisig.rs]: Leftover `println!(...)`

File: rust/chains/hyperlane-cardano/src/rpc/mod.rs [hunk 13]
```diff
@@ -51,6 +55,100 @@ impl CardanoRpc {
...
+        let checkpoint_root = hex::encode(&metadata[32..64]);
+        let checkpoint_index = u32::from_be_bytes(metadata[64..68].try_into().unwrap());
+        let signatures = metadata[68..metadata.len() - 1]
+            .chunks(64)
+            .map(|s| hex::encode(s))
+            .collect();
+        estimate_inbound_message_fee(
+            &self.0,
+            EstimateInboundMessageFeeRequest {
+                relayer_cardano_address:
```
Comment (serejke) [inline rust/chains/hyperlane-cardano/src/rpc/mod.rs]: I'd suggest moving the relayer private key and address configs to RPC ENV variables.

In the current architecture, RPC service is run by relayer operator and is trustful.

We don't have capacity for full Ed25519 signing support on relayer, but can later address that in https://github.com/tvl-labs/hyperlane-cardano/issues/51
